packages/tanstack-react-query/test/queryOptions.test.tsx

```diff
@@ -0,0 +1,356 @@
+import { getServerAndReactClient } from './__helpers';
+import { skipToken, useQuery, useSuspenseQuery } from '@tanstack/react-query';
+import { render, waitFor } from '@testing-library/react';
+import { initTRPC } from '@trpc/server';
+import { createDeferred } from '@trpc/server/unstable-core-do-not-import';
+import { konn } from 'konn';
+import * as React from 'react';
+import { describe, expect, expectTypeOf, test, vi } from 'vitest';
+import { z } from 'zod';
+
+const ctx = konn()
+  .beforeEach(() => {
+    let iterableDeferred = createDeferred<void>();
+    const nextIterable = () => {
+      iterableDeferred.resolve();
+      iterableDeferred = createDeferred();
+    };
+    const t = initTRPC.create({});
+
+    const appRouter = t.router({
+      post: t.router({
+        byId: t.procedure
+          .input(
+            z.object({
+              id: z.string(),
+            }),
+          )
+          .query(() => '__result' as const),
+        byIdWithSerializable: t.procedure
+          .input(
+            z.object({
+              id: z.string(),
+            }),
+          )
+          .query(() => ({
+            id: 1,
+            date: new Date(),
+          })),
+        iterable: t.procedure.query(async function* () {
+          for (let i = 0; i < 3; i++) {
+            await iterableDeferred.promise;
+            yield i + 1;
+          }
+        }),
+      }),
+    });
+
+    const testHelpers = getServerAndReactClient(appRouter);
+
+    return {
+      ...testHelpers,
+      nextIterable,
+    };
+  })
+  .afterEach(async (ctx) => {
+    await ctx?.close?.();
+  })
+  .done();
+
+describe('queryOptions', () => {
+  test('basic', async () => {
+    const { App, useTRPC } = ctx;
+    function MyComponent() {
+      const trpc = useTRPC();
+      const queryOptions = trpc.post.byId.queryOptions({ id: '1' });
+      expect(queryOptions.trpc.path).toBe('post.byId');
+      const query1 = useQuery(queryOptions);
+
+      const query2 = useQuery(trpc.post.byId.queryOptions({ id: '1' }));
+      expectTypeOf(query1).toMatchTypeOf(query2);
+
+      if (!query1.data) {
+        return <>...</>;
+      }
+
+      expectTypeOf(query1.data).toMatchTypeOf<'__result'>();
+
+      return <pre>{JSON.stringify(query1.data ?? 'n/a', null, 4)}</pre>;
+    }
+
+    const utils = render(
+      <App>
+        <MyComponent />
+      </App>,
+    );
+    await waitFor(() => {
+      expect(utils.container).toHaveTextContent(`__result`);
+    });
+  });
+
+  test('select', async () => {
+    const { useTRPC, App } = ctx;
+    function MyComponent() {
+      const trpc = useTRPC();
```

Comment:
my only question is if we could use `use()` to get the TrpcContext without needing to useTRPC everywhere, just with a singleton reference to the proxy

In any case I love it.

Reply:
it's really not a big deal to useTRPC anyway, just a dream

Reply:
i think we could, but we don't have React 19 on the project, it's not a big deal either way as you'd need a `useQuery()` for any data fetching

tried to upgrade in #6184, but getting lots of errors